### PR TITLE
Add recipe for C++-only Protocol Buffers to shadow repo.continuum.io libprotobuf

### PR DESCRIPTION
**Pull request by wesm**
Closes https://github.com/conda-forge/protobuf-feedstock/issues/25

Having the C++ libraries depend on Python was a rough edge that I encountered building a C++ project with conda-forge libraries

### CODE REVIEW

File: recipes/libprotobuf/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  url: https://github.com/google/protobuf/archive/v{{ version }}/protobuf-v{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  features:
+    - vc9                # [win and py27]
+    - vc10               # [win and py34]
+    - vc14               # [win and py35]
+    - vc14               # [win and py36]
```
Comment (isuruf) [inline recipes/libprotobuf/meta.yaml]: You can skip one of `py35` and `py36`

File: recipes/libprotobuf/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    - git
+    - cmake              # [win]
+    - autoconf           # [unix]
+    - automake           # [unix]
+    - libtool            # [unix]
+    - pkg-config         # [unix]
+    - zlib 1.2.*
+    - python             # [win]
+  run:
+    - zlib 1.2.*
```
Comment (isuruf) [inline recipes/libprotobuf/meta.yaml]: Add following as mentioned here, http://conda-forge.github.io/docs/buildwin.html#vc-features
```
    - vc 9  # [win and py27]
    - vc 10  # [win and py34]
    - vc 14  # [win and py>=35]
```

File: recipes/libprotobuf/build.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+if [ "$(uname)" == "Darwin" ];
+then
+    # Switch to clang with C++11 ASAP.
+    export MACOSX_VERSION_MIN=10.7
+    export CC=clang
+    export CXX=clang++
+    export CXXFLAGS="-mmacosx-version-min=${MACOSX_VERSION_MIN}"
+    export CXXFLAGS="${CXXFLAGS} -stdlib=libc++ -std=c++11 -O2"
+    export LIBS="-lc++"
```
Comment (isuruf) [inline recipes/libprotobuf/build.sh]: This has deviated from https://github.com/conda-forge/toolchain-feedstock/blob/master/recipe/activate.sh
I'd suggest using toolchain, but some people prefer having it explicitly in the recipe.

File: recipes/libprotobuf/build.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+./configure --prefix="${PREFIX}" \
+            --with-pic \
+            --enable-shared \
+            --enable-static \
+	    CC="${CC}" \
+	    CXX="${CXX}" \
+	    CXXFLAGS="${CXXFLAGS}" \
+	    LDFLAGS="${LDFLAGS}"
+make
```
Comment (isuruf) [inline recipes/libprotobuf/build.sh]: Change this to 
```
make -j${CPU_COUNT}
make check -j${CPU_COUNT}
```
for faster compilation

File: recipes/libprotobuf/bld.bat [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+:: Configure and install based on protobuf's instructions and other `bld.bat`s.
+cmake -G "NMake Makefiles" ^
+         -DCMAKE_BUILD_TYPE=Release ^
+         -DCMAKE_PREFIX_PATH=%LIBRARY_PREFIX% ^
+         -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
+         -Dprotobuf_WITH_ZLIB=ON ^
+         ../..
+if errorlevel 1 exit 1
+nmake
```
Comment (isuruf) [inline recipes/libprotobuf/bld.bat]: If you used `ninja`, you can build in parallel here as well.

File: recipes/libprotobuf/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+{% set name = "libprotobuf" %}
```
Comment (jakirkham) [inline recipes/libprotobuf/meta.yaml]: Is this the name they are using in `defaults`?

File: recipes/libprotobuf/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  commands:
+    - protoc --help
+    - test -f ${PREFIX}/lib/libprotobuf.a                           # [unix]
+    - test -f ${PREFIX}/lib/libprotobuf.dylib                       # [osx]
+    - test -f ${PREFIX}/lib/libprotobuf.so                          # [linux]
+    - if not exist %PREFIX%\\Library\\lib\\libprotobuf.lib exit 1   # [win]
+
+about:
+  home: https://developers.google.com/protocol-buffers/
+  license: BSD 3-Clause
```
Comment (jakirkham) [inline recipes/libprotobuf/meta.yaml]: I know we didn't have it before, but we should add `license_file: LICENSE` to package the license file from the source code.

File: recipes/libprotobuf/meta.yaml (hunk 6 above)
Comment (wesm) [inline recipes/libprotobuf/meta.yaml]: Yes

File: recipes/libprotobuf/build.sh [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+
+if [ "$(uname)" == "Darwin" ];
+then
+    # Switch to clang with C++11 ASAP.
+    export MACOSX_VERSION_MIN=10.7
+    export CC=clang
+    export CXX=clang++
+    export CXXFLAGS="-mmacosx-version-min=${MACOSX_VERSION_MIN}"
+    export CXXFLAGS="${CXXFLAGS} -stdlib=libc++ -std=c++11 -O2"
```
Comment (jakirkham) [inline recipes/libprotobuf/build.sh]: Also same story with the `-O2` flag. This should move out of these `if` statements.

**Comment by dougalsutherland** [comment]
This would be useful for me as well.

**Comment by wesm** [comment]
Got tied up. Can return to this recipe next week unless someone wants to take it over

**Comment by jakirkham** [comment]
cc @hajapy

**Comment by hajapy** [comment]
In principle I am in favour of this but I just want to ensure we will be able to build the python package from libprotobuf. Has anyone tried this?

Also, this recipe/meta has diverged a bit from the last successful recipe/meta for the python package. Please have a look at the simplifications introduced during https://github.com/conda-forge/protobuf-feedstock/pull/29 or just look at the latest master branch.

**Comment by jakirkham** [comment]
Friendly nudge. 😉 Would be nice to revive this.

**Comment by jakirkham** [comment]
Would recommend refreshing this from the feedstock. We are at 3.5.0 currently.

File: recipes/libprotobuf/build.sh (hunk 3 above)
Comment (jakirkham) [reply recipes/libprotobuf/build.sh]: Added PR ( https://github.com/conda-forge/protobuf-feedstock/pull/44 ) to correct `protobuf`.

**Comment by jakirkham** [comment]
Have pushed some commits to copy the content over from `protobuf` and addresses the review comments above. This could use a review, @conda-forge/staged-recipes. Would like to get this in so that we can update `conda-forge-pinning` to use `libprotobuf` instead of `protobuf`.

**Comment by dougalsutherland** [comment]
GPy failure on Circle looks like maybe this should be rebased? GPy was merged in June (https://github.com/conda-forge/staged-recipes/pull/3084).

Otherwise, looks good to me, though it'd be useful to make sure that we can build the python bindings off of this successfully (https://github.com/conda-forge/staged-recipes/pull/3087#issuecomment-316867301).

**Comment by jakirkham** [comment]
Yeah, there were some changes to the CircleCI configuration at `staged-recipes` since this PR was opened. Thanks for the reminder about that. Have merged with current `master`, which should clear that error. (GPy is gone from my local copy after merging)

**Comment by jakirkham** [comment]
> ...though it'd be useful to make sure that we can build the python bindings off of this successfully (https://github.com/conda-forge/staged-recipes/pull/3087#issuecomment-316867301).

Sure. I pushed a commit that includes a prospective `protobuf` recipe that builds off of this `libprotobuf`. Have tested it locally on one Python combination on my Mac and it works fine, but we can see if that holds true for the CIs. We should revert that commit once we are satisfied before merging.

**Comment by wesm** [comment]
cc @MaxRis

**Comment by jakirkham** [comment]
The build with the prototype Python-only `protobuf` package using `libprotobuf` passed above (excepting the linter of course). Have pushed a commit to revert it so this PR only contains `libprotobuf`. Have skipped CI in the last commit as this has already been tested. Thinking this is good to go now. Final comments before merging @conda-forge/staged-recipes?

**Comment by scopatz** [comment]
LGTM

**Comment by wesm** [comment]
thank you!

**Comment by jakirkham** [comment]
Thanks for getting this started. :)

**Comment by jakirkham** [comment]
Updating `protobuf` to depend on `libprotobuf` in PR ( https://github.com/conda-forge/protobuf-feedstock/pull/45 ). Please take a look if you have time.

**Pull request merged**